File: scripts/check_q5_0.py
#!/usr/bin/env python3
"""Reference side of the Q5_0 dequant cross-check (see check_q5_0.mjs).

    python3 scripts/check_q5_0.py <model.gguf> <tensor-name> [nElems]

Uses the `gguf` package's GGUFReader + dequantize — the llama.cpp-blessed
reference implementation. Prints one value per line.
"""
import sys

import numpy as np
from gguf import GGUFReader
from gguf.quants import dequantize

file, name = sys.argv[1], sys.argv[2]
n = int(sys.argv[3]) if len(sys.argv) > 3 else 64

reader = GGUFReader(file)
t = next(t for t in reader.tensors if t.name == name)
print(f"tensor {t.name} type={t.tensor_type} shape={t.shape}", file=sys.stderr)

vals = dequantize(t.data, t.tensor_type).flatten()
for v in vals[:n]:
    print(f"{v:.8g}")

File: scripts/check_q5_0.mjs
// Cross-check the src/gguf.ts Q5_0 dequant against the python `gguf` reference.
//
//   node --experimental-strip-types scripts/check_q5_0.mjs <model.gguf> <tensor> [nElems]
//
// Imports src/gguf.ts directly (Node ≥22.6 type stripping; the only import in
// gguf.ts is type-only, so no further resolution is needed). Reads the header
// from a 64 MB prefix, then reads ONLY the target tensor's bytes and dequants.
// Prints one value per line for diffing against scripts/check_q5_0.py.
import { openSync, readSync } from 'node:fs';
import { GGUFParser, tensorByteSize } from '../src/gguf.ts';

const [file, tensorName, nStr] = process.argv.slice(2);
const N = Number(nStr ?? 64);

const HEAD = 64 * 1024 * 1024;
const fd = openSync(file, 'r');
const head = Buffer.alloc(HEAD);
readSync(fd, head, 0, HEAD, 0);
const gguf = new GGUFParser(head.buffer.slice(head.byteOffset, head.byteOffset + HEAD)).parse();

const t = gguf.tensors.find((x) => x.name === tensorName);
if (!t) { console.error(`tensor ${tensorName} not found`); process.exit(1); }
console.error(`tensor ${t.name} type=${t.type} dims=${t.dims} offset=${t.offset} dataOffset=${gguf.dataOffset}`);

const nbytes = tensorByteSize(t);
const buf = Buffer.alloc(nbytes);
readSync(fd, buf, 0, nbytes, gguf.dataOffset + Number(t.offset));
const tp = new GGUFParser(buf.buffer.slice(buf.byteOffset, buf.byteOffset + nbytes));
const vals = tp.getTensorData({ ...t, offset: 0n }, 0);

console.log(Array.from(vals.slice(0, N)).map((v) => v.toPrecision(8)).join('\n'));
